python/ray/data/datasource/image_folder_datasource.py

```diff
@@ -133,6 +152,97 @@ def _read_file(
+    def _set_encoding_ratio(self, encoding_ratio: int):
+        """Set image file encoding ratio, to provide accurate size in bytes metadata."""
+        self._encoding_ratio = encoding_ratio
+
+    def _get_block_metadata(
+        self,
+        paths: List[str],
+        schema: Optional[Union[type, "pyarrow.lib.Schema"]],
+        *,
+        rows_per_file: Optional[int],
+        file_sizes: List[Optional[int]],
+    ) -> BlockMetadata:
+        metadata = super()._get_block_metadata(
+            paths, schema, rows_per_file=rows_per_file, file_sizes=file_sizes
+        )
+        if metadata.size_bytes is not None:
+            metadata.size_bytes = int(metadata.size_bytes * self._encoding_ratio)
+        return metadata
+
+
+class _ImageFolderDatasourceReader(_FileBasedDatasourceReader):
+    def __init__(
+        self,
+        delegate: FileBasedDatasource,
+        paths: List[str],
+        filesystem: "pyarrow.fs.FileSystem",
+        partition_filter: PathPartitionFilter,
+        meta_provider: _ImageFileMetadataProvider = _ImageFileMetadataProvider(),
+        **reader_args,
+    ):
+        super().__init__(
+            delegate=delegate,
+            paths=paths,
+            filesystem=filesystem,
+            schema=None,
+            open_stream_args=None,
+            meta_provider=meta_provider,
+            partition_filter=partition_filter,
+            **reader_args,
+        )
+        self._encoding_ratio = self._estimate_files_encoding_ratio()
+        meta_provider._set_encoding_ratio(self._encoding_ratio)
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        return sum(self._file_sizes) * self._encoding_ratio
+
+    def _estimate_files_encoding_ratio(self) -> float:
+        """Return an estimate of the image files encoding ratio."""
+        # Prefetch one image file into memory, and use it to estimate data size for
+        # all images, because all images are homogeneous with same size after resizing.
+        import pandas as pd
+
+        start_time = time.perf_counter()
+
+        # Filter out empty file to avoid noise.
+        non_empty_path_and_size = list(
+            filter(lambda p: p[1] > 0, zip(self._paths, self._file_sizes))
+        )
+        if len(non_empty_path_and_size) == 0:
+            logger.warn(
+                "All input image files are empty. "
+                "Use on-disk file size to estimate images in-memory size."
+            )
+            return IMAGE_ENCODING_RATIO_ESTIMATE_DEFAULT
+
+        # Prefetch first non-empty image file to do estimation.
+        path = non_empty_path_and_size[0][0]
+        single_image_block: pd.DataFrame = self._delegate._read_file(
+            self._filesystem.open_input_stream(path),
+            path,
+            self._reader_args.get("root"),
+            self._reader_args.get("size"),
+        )
+        single_image_memory_size = single_image_block.memory_usage().sum()
```

Comment:
Don't we need `deep=True` here? https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.memory_usage.html

Reply:
@clarkzinzow - good catch, I think we should, let me add it.

Reply:
@clarkzinzow - added.